### ISSUE TITLE
Advice about the v8go API

### ISSUE THREAD

**Issue by kuoruan** [rogchap/v8go#80]
Currently, the `Context` type in v8go hold the `v8:Context` ptr of v8, but Golang already has a type `context.Context`

Because of this, we must write more code to impl timeout.

https://github.com/rogchap/v8go#terminate-long-running-scripts

So, we can just hide the `v8:Context`.

To create a new v8go object, use

```
v8, _ := v8go.New()
```

or

```
iso, _ := v8go.NewIsolate()
v8, _ := v8go.New(iso)
```

To run a script with timeout

```
ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
defer cancel()

v8.RunScriptContext(ctx, "const add = (a, b) => a + b", "math.js")
```

Like other Golang library do.

**Comment by rogchap** [comment]
Thanks for the suggestion @kuoruan 

I agree that the use of the work `Context` is conflicting, but I guess that is why we have package scopes in Go.

One of the goals of the project is to keep the API as close to the C++ API as possible; so would be inclined to keep the idea of the v8 Context.

That said, there is a second part to the goals that state: "idiomatic Go"; which `context.Context` is very much part of; which makes it a conflicting argument 🤔 

Once we get closer to a stable API, I'll definitely look at how we can incorporate `context.Context`; it would definitely be of benefit, especially for some of the callback logic.

**Comment by mitar** [comment]
+1 on integrating `context.Context`.

**Comment by dylanahsmith** [comment]
I don't think it makes sense to hide the v8 context, since it has a broader purpose than context.Context.

Also, I think supporting the full context capabilities internally to v8go would effectively mean starting a goroutine in order to be able to handle a cancel event.  If that needs to be added on each v8go call that could invoke JS execution, then that could be a lot of starting and stopping of goroutines.  

Note that JS execution can happen in a lot more APIs than just running scripts.  For instance, even the value casting functions could trigger JS execution.  E.g. v8::Value::ToString could call the [Symbol.toPrimitive method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toPrimitive)).  It would add a lot of API surface to provide alternative versions of all of these methods to handle a context.

An alternative would be to just support setting and clearing deadline, which could be done using separate methods.  This would also be cheaper, since it could just be done using a timer created using `time.AfterFunc`, which could be stopped or reset when the deadline is cleared or set a following time.  If unused, the field could just be left `nil`.   However, unlike TerminateExecution, we would probably want to prevent further JS execution using a boolean flag, in case TerminateExecution gets called outside of the JS code (e.g. between setting the deadline and starting JS execution or between calls that trigger JS execution).

The other thing to note is that forcing JS execution at an arbitrary point in the JS code can leave the V8 context is an unreliable state.  For instance, if the JS code sets some global variable and tries to reset it in a finally block, then I think that finally block code could end up being skipped.  It may be worth considering whether the feature will be supporting an anti-pattern over starting with a more graceful timeout mechanism and only using a more forceful timeout after a delay.  Nothing new is needed for either approach, so anything added would just be making some approach(es) more convenient.

**Comment by molon** [comment]
```
func WrapRun(
	ctx context.Context,
	v8ctx *v8go.Context,
	f func(v8ctx *v8go.Context) (*v8go.Value, error),
) (result *v8go.Value, terminated bool, err error) {
	if ctx.Err() != nil {
		return nil, false, ctx.Err()
	}

	stop := context.AfterFunc(ctx, func() {
		v8ctx.Isolate().TerminateExecution()
	})
	defer func() {
		if !stop() {
			terminated = true
		}
	}()

	result, err = f(v8ctx)
	if err != nil {
		if ctx.Err() != nil {
			return nil, false, ctx.Err()
		}
		return nil, false, err
	}
	return result, false, nil
}

```
